Add `Unsupported` to `std::io::ErrorKind`

CDirkx opened the pull request:
I noticed a significant portion of the uses of `ErrorKind::Other` in std is for unsupported operations.
The notion that a specific operation is not available on a target (and will thus never succeed) seems semantically distinct enough from just "an unspecified error occurred", which is why I am proposing to add the variant `Unsupported` to `std::io::ErrorKind`. 

**Implementation**:

The following variant will be added to `std::io::ErrorKind`:
 
```rust
/// This operation is unsupported on this platform.
Unsupported
```
`std::io::ErrorKind::Unsupported` is an error returned when a given operation is not supported on a platform, and will thus never succeed; there is no way for the software to recover. It will be used instead of `Other` where appropriate, e.g. on wasm for file and network operations. 

`decode_error_kind` will be updated  to decode operating system errors to `Unsupported`:
- Unix and VxWorks: `libc::ENOSYS`
- Windows: `c::ERROR_CALL_NOT_IMPLEMENTED`
- WASI: `wasi::ERRNO_NOSYS`

**Stability**:
This changes the kind of error returned by some functions on some platforms, which I think is not covered by the stability guarantees of the std? User code could depend on this behavior, expecting `ErrorKind::Other`, however the docs already mention:

> Errors that are `Other` now may move to a different or a new `ErrorKind` variant in the future. It is not recommended to match an error against `Other` and to expect any additional characteristics, e.g., a specific `Error::raw_os_error` return value.

The most recent variant added to `ErrorKind` was `UnexpectedEof` in `1.6.0` (almost 5 years ago), but `ErrorKind` is marked as `#[non_exhaustive]` and the docs warn about exhaustively matching on it, so adding a new variant per se should not be a breaking change.

The variant `Unsupported` itself could be marked as `#[unstable]`, however, because this PR also immediately uses this new variant and changes the errors returned by functions I'm inclined to agree with the others in this thread that the variant should be insta-stabilized.

rust-highfive commented:
r? @shepmaster

(rust_highfive has picked a reviewer for you, use r? to override)

jyn514 commented on library/std/src/sys/unsupported/os.rs:
```diff
@@ -80,11 +80,11 @@ pub fn getenv(_: &OsStr) -> io::Result<Option<OsString>> {
 
 pub fn unsetenv(_: &OsStr) -> io::Result<()> {
-    Err(io::Error::new(io::ErrorKind::Other, "cannot unset env vars on this platform"))
+    Err(io::Error::new(io::ErrorKind::NotSupoorted, "cannot unset env vars on this platform"))
```
```suggestion
    Err(io::Error::new(io::ErrorKind::NotSupported, "cannot unset env vars on this platform"))
```

scottmcm commented:
cc @rust-lang/libs explicitly for this -- this might need to go in insta-stable, since while the variant can be marked unstable, the change to stop emitting the previous one would not be feature-gated.

the8472 commented:
Should `decode_error_kind` be extended to translate any of `libc::{ENOSYS,ENOTSUP,EOPNOTSUPP}` to this ErrorKind?

CDirkx commented:
I updated `decode_error_kind` to support `NotSupported` on various platforms:

- Unix and VxWorks: `libc::{ENOSYS, ENOTSUP, EOPNOTSUPP}`
- Windows: `c::{ERROR_CALL_NOT_IMPLEMENTED, E_NOTIMPL}`
- CloudABI: `abi::errno::{NOSYS, NOTSUP}` (has no `OPNOTSUPP`)
- WASI: `wasi::{ERRNO_NOSYS, ERRNO_NOSUP}` (has no `OPNOTSUPP`)
- Hermit: (has hard coded error numbers, I could not find a relevant error number)
- SGX: (has no equivalent error in `fortanix_sgx_abi::Error`)

I didn't include errors like `EAFNOSUPPORT` (address family not supported) and `EPROTONOSUPPORT` (protocol not supported) for now.

m-ou-se commented:
> user code could depend on this behavior, expecting `ErrorKind::Other`

The documentation fortunately already notes that these kind of changes might happen in the future:

> Errors that are `Other` now may move to a different or a new `ErrorKind` variant in the future. It is not recommended to match an error against `Other` and to expect any additional characteristics, e.g., a specific `Error::raw_os_error` return value.

(In hindsight it might have been better if matching 'other' errors against `ErrorKind::Other` was not possible. E.g. by using an unstable/doc(hidden) variant for those. But I guess it's too late for that now.)

I'm not sure if adding  an `#[unstable]` variant to this stable enum is a good idea. Especially as these errors used a `#[stable]` variant before, like @scottmcm mentioned.

> CloudABI: `abi::errno::{NOSYS, NOTSUP}` (has no `OPNOTSUPP`)

Feel free to ignore CloudABI, as that platform itself will be NOTSUP soon: https://github.com/rust-lang/rust/pull/78439

KodrAus commented:
I can see the value in wanting to check whether or not an IO operation is supported so you can fallback to alternatives.

I also think we should consider marking this as insta-stable if we want to add it. In either case I think an FCP would probably be a good idea before we land it.

CDirkx commented:
I have updated the PR to immediately stabilize `NotSupported` and updated the description to include the information inthis thread.

shepmaster commented:
r? @KodrAus

bors commented:
:umbrella: The latest upstream changes (presumably #79060) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

rust-log-analyzer commented:
The job `x86_64-gnu-tools` of your PR [failed](https://github.com/rust-lang/rust/runs/1428687958) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang/rust/commit/1918e9a8a32ccc1d31cfb6fe0b2b8f2b460cef75/checks/1428687958/logs), [raw log](https://github.com/rust-lang/rust/commit/1918e9a8a32ccc1d31cfb6fe0b2b8f2b460cef75/checks/1428687958/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[command]/usr/bin/git submodule foreach --recursive git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :
[command]/usr/bin/git config --local http.https://github.com/.extraheader AUTHORIZATION: basic ***
##[endgroup]
##[group]Fetching the repository
[command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=2 origin +965b423fc46ef69fcbfa3a5fb6363ad3f49a7279:refs/remotes/pull/78880/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-lang/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-lang/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

KodrAus commented:
@rfcbot fcp merge

This proposes adding a `NotSupported` variant to `io::ErrorKind` that's used when an operation isn't supported. Callers may want to try falling back to alternatives or bubble up.

KodrAus commented on library/std/src/io/error.rs:
```diff
@@ -178,6 +178,10 @@ pub enum ErrorKind {
     UnexpectedEof,
+
+    /// This operation is not supported on this platform.
+    #[stable(feature = "not_supported_error", since = "1.50.0")]
```
Just a note that we'll need to bump this `since` field if the FCP completes

BurntSushi commented:
Could we do a crater run with this change? I know this change is technically okay, but I think it would still be a good idea to get an understanding of what kind of breakage this might cause, if any.

jyn514 commented:
@bors try

(`try` doesn't take much time even if we decide not to do a crater run)

bors commented:
:hourglass: Trying commit dd626891bc56b6fbd696c24ae62e46148ca30040 with merge 52e876ecd5b4d95f3876e59d428e473441aa4220...
<!-- homu: {"type":"TryBuildStarted","head_sha":"dd626891bc56b6fbd696c24ae62e46148ca30040","merge_sha":"52e876ecd5b4d95f3876e59d428e473441aa4220"} -->

Mark-Simulacrum commented:
> Could we do a crater run with this change? I know this change is technically okay, but I think it would still be a good idea to get an understanding of what kind of breakage this might cause, if any.

We can, but it seems unlikely to turn anything up. Crater runs with modern linux kernels (5.3 and 5.4 on the two builders I checked), and so it seems unlikely that the test results will have any major cases of unsupported operations.

BurntSushi commented:
@Mark-Simulacrum Yeah, I was thinking of case analysis failures. But it looks like this enum is tagged with `#[non_exhaustive]`, so yeah, probably won't turn anything up.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1705843834)
Build commit: 52e876ecd5b4d95f3876e59d428e473441aa4220 (`52e876ecd5b4d95f3876e59d428e473441aa4220`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1705843834"},"merge_sha":"52e876ecd5b4d95f3876e59d428e473441aa4220"} -->

ijackson commented:
I approve of adding more ErrorKinds, so sorry to throw a spanner in the works here.  But I don't think the set of errno values on Unix is right.

Also in general I question whether it is a good idea to bundle different errno values together into one ErrorKind.  Experienced Unix system administrators recognise errno values and the corresponding strings and can often infer useful information from quite subtle distinctions.

In detail:

 * `ENOSYS` means specifically that the system call being called is not implemented.  In practice this probably means that the executable is running on too-old a kernel, or that there is some kind of libc incompatibility.  `ENOSYS` is not typically used for situations when a specific system call parameter value is not supported.
 * `EOPNOTSUPP` means specifically that a socket operation was not supporrted.  This much more specific error generally means that the sequence of system calls on a socket was not appropriate; for example, calling `accept` on a connected socket.  But it can mean other specific things depending on what precisely the fd refers to.

Note that there are a lot of other errno values that refer to more specific kinds of unsupportedness.

I am OK with merging `EOPNOTSUP` and `ENOTSUP` (SuS says they may be the same).

I think for this MR the best answer is probably to include only `ENOSYS` and leave other errno values for future extensions to `ErrorKind`.  This narrowing may warrant a slight change to the description.

the8472 commented:
> Experienced Unix system administrators recognise errno values and the corresponding strings and can often infer useful information from quite subtle distinctions.

`ErrorKind` represents a cross-platform rust side representation of an error that is useful to `match` on. The errno is not discarded, you can query and print it directly if you want more fine-grained, platform-specific information.

ijackson commented:
> `ErrorKind` represents a cross-platform rust side representation of an error that is useful to `match` on. The errno is not discarded, you can query and print it directly if you want more fine-grained, platform-specific information.

Yes.  But smooshing together errors which can be usefully distinguished, can make it harder to write portable code which knows a lot about the details of a particular OS API area.  This is particularly true of stuff relating to sockets and networking, where the influence of the BSD socket API and therefore of Unix errno values is strong.

I would prefer to avoid asking people to write code which, for each platform, imports `os::...`, uses `into_raw`, and matches on whatever that platform's spelling is for the socket error `EOPNOTSUPP`.  If we provide a distinct `ErrorKind` for that, we would save them that effort.  And we have a good place to put that code.

ijackson commented:
I guess at this point I should mention https://github.com/rust-lang/rust/pull/79965 which is still on my todo list to clean up and resubmit with Windows code alongside the Unix code.

the8472 commented:
Do you have a motivating example where ENOSYS and EOPNOTSUPP can both occur on the same syscall and would require different handling?

ijackson commented:
If you call `connect` and get `ENOSYS` it means that you are using the wrong syscall ABI, or the wrong binary has been installed for the platform.  Maybe the right answer is to exec a binary for a different kernel ABI.  If you get `EOPNOTSUPP` it means that the fd you are calling it on is not in the state you expected.  That might be an expected condition (eg, the socket is already connected).

ijackson commented:
If you call `fchmodat` with `AT_SYMLINK_NOFOLLOW` and get `ENOSYS` you must fall back to some dance with `chdir` or something.  If you get `EOPNOTSUPP` it means the target was a symlink when you expected it not to be (probably for security reasons)  and the OS doesn't support symlink modesand the dance (if you went through it) would fail too.

m-ou-se commented:
@rfcbot concern EOPNOTSUPP

See discusson starting at https://github.com/rust-lang/rust/pull/78880#issuecomment-777623233

ijackson commented:
I am no expert on the wasi environmen but I would guess that WASI `ERRNO_NOSUP` is `EOPNOTSUPP` with a funny hat on.  In particular, that you get it from unexpected sequences of socket operations.

wasi is an environment where compatibility with different external environments, and fallback strategies, is more important than for native binaries.  So we should definitely try to DTRT there.

The best spec I found for what errno values there are, was this https://github.com/WebAssembly/WASI/blob/main/phases/snapshot/docs.md which bears a striking resemblance to a Unix errno list.

So probably whatever we do with `EOPNOTSUPP` (and `ENOTSUP`) should be done for wasi `ERRNO_NOSUP` too.

the8472 commented:
I'm not convinced by the given examples.

> If you call `connect` and get `ENOSYS` it means that you are using the wrong syscall ABI, or the wrong binary has been installed for the platform. Maybe the right answer is to exec a binary for a different kernel ABI. If you get `EOPNOTSUPP` it means that the fd you are calling it on is not in the state you the call expected. That might be an expected condition (eg, the socket is already connected).

Perhaps just ignorance on my end, but when would you expect to be able to run a program well enough to the point where you hit the wrong kernel ABI and then somehow recover from that problem? This doesn't seem like a recoverable thing you would deal with from within the program. You wouldn't even be able to print an error message or call exit/abort if that is misconfigured.

> If you call `fchmodat` with `AT_SYMLINK_NOFOLLOW` and get `ENOSYS` you must fall back to some dance with `chdir` or something. If you get `EOPNOTSUPP` it means the target was a symlink when you expected it not to be (probably for security reasons) and the OS doesn't support symlink modes and the dance (if you went through it) would fail too.

Does that apply to any supported rust platforms? At least on linux `fchmodat` was inttroduced in 2.6.16, which is older than the minimum supported version 2.6.32.

ijackson commented:
> > If you call `connect` and get `ENOSYS` it means that you are using the wrong syscall ABI, or the wrong binary has been installed for the platform. Maybe the right answer is to exec a binary for a different kernel ABI. If you get `EOPNOTSUPP` it means that the fd you are calling it on is not in the state you the call expected. That might be an expected condition (eg, the socket is already connected).
> 
> Perhaps just ignorance on my end, but when would you expect to be able to run a program well enough to the point where you hit the wrong kernel ABI and then somehow recover from that problem?

There have in the past been kernel APIs where subsets of the system calls have been replaced with newer variants.  For example, large file support generally involved replacing calls to `fstat` et al.  You can see this as `lstat64` in strace on Linux.  Our API choices about errno handling are not just for now; they are for the future - both the future of majority platforms like Linux, and the future of Rust which we hope will become more widely ported.

In practice I agree that the most correct thing to do in most cases on getting `ENOSYS` is going to be to bail completely.  But even so in a program that is doing exciting socket stunts, that is a different appropriate response to the one for `EOPNOTSUPP`, which might well be to draw conclusions about the socket state and respond by making different socket calls.

Even in a very program, where the errors are handled by printing an error message and quitting, the error messages are different.  `ENOSYS` probably indicates a version incompatibility between the program and its environment, for which a message saying something about "unsupported" is appropriate.  The appropriate response of the system administrator is to double check the installation approach.

In a simple program, `EOPNOTSUPP` probably indicates a logical error in the program: it has made its socket calls in the wrong order, or the wrong combination, perhaps provoked by an unusual set of command line options or configuration settings.  The appropriate response for the user is probably to try to run the program a different way or to ask it to do something more simple.  The appropriate response for the programmer to the bug report is to look at the sequence of socket calls and fix the program's socket handling logic.

> > If you call `fchmodat` with `AT_SYMLINK_NOFOLLOW` and get `ENOSYS` you must fall back to some dance with `chdir` or something. If you get `EOPNOTSUPP` it means the target was a symlink when you expected it not to be (probably for security reasons) and the OS doesn't support symlink modes and the dance (if you went through it) would fail too.
> 
> Does that apply to any supported rust platforms? At least on linux `fchmodat` was inttroduced in 2.6.16, which is older than the minimum supported version 2.6.32.

I don't think we want to rule out supporting Rust on platforms without `fchmodat`!  So reasoning like this from the set of currently-supported platforms, to the proper treatment of Unix errnos forever more for all Rust programs, is not correct.

the8472 commented:
The examples haven't been particularly convincing but now I agree that there is a categorical distinction between those two regardless of current practical impact. The distinction is whether something is unconditionally unavailable (`ENOSYS`) or merely not applicable to a particular resource or in a particular state (`EOPNOTSUPP`). Although `EINVAL` is frequently used for the latter purpose too, e.g. when you pass the wrong kind of file descriptor to a syscall for example.

The doc comment already reflects this.

```rust
/// This operation is not supported on this platform.
```

ijackson commented:
@the8472  Let us try to move forward; I'm not enjoying this debate very much and I guess neither are you :-/.

In practical terms for this MR, I think your last comment means you agree that `EOPNOTSUPP`, `ENOTSUP` and WASI's `ERRNO_NOSUP` ought to be dropped from the list of errno values for the new `ErrorKind::Unimplemented`?

the8472 commented:
> In practical terms for this MR, I think your last comment means you agree that `EOPNOTSUPP`, `ENOTSUP` and WASI's `ERRNO_NOSUP` ought to be dropped from the list of errno values for the new `ErrorKind::Unimplemented`?

Yes.

CDirkx commented:
Looking deeper into the semantics of `EOPNOTSUPP` and `ENOTSUP`, I agree with the distinction from `ENOSYS`. The original motivation for including those error codes wasn't that strong anyway, and I am glad that people with more experience with UNIX errors than me weighed in.

To summarize the specific semantics in this proposal:
`std::io::ErrorKind::NotSupported` is an error returned when a given operation is not supported on a platform, and will thus never succeed; there is no way for the software to recover.

The only error codes that will map to `NotSupported`:
 - UNIX: `ENOSYS` "The return of [ENOSYS] is to be taken to indicate that the function of the interface is not supported at all; the function will always fail with this error code." ([POSIX.1-2001](http://www.open-std.org/jtc1/sc22/open/n4217.pdf#page=3452))
 - Windows: `ERROR_CALL_NOT_IMPLEMENTED` "This function is not supported on this system." ([Windows System Error Codes](https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-))

That means I will drop the originally included error codes `EOPNOTSUPP` `ENOTSUP` `ERRNO_NOSUP` and `E_NOTIMPL`, as they all have meanings that do not directly correspond to `NotSupported`.

m-ou-se commented:
@rfcbot resolve EOPNOTSUPP

CDirkx commented:
I apparently missed some places `NotSupported` can be used, I added a commit to also include those.
I also updated `since` so this should now be ready to merge.

joshtriplett commented:
I only just now saw this. I realize that it has already been through FCP, but I'm concerned about the name, specifically.

I don't think this is the right name for ENOSYS. It's already challenging, sometimes, to map these names to the underlying errno code. This name, in particular, seems much more likely to mean EOPNOTSUPP rather than ENOSYS, even though the semantic usage is like the latter, and Rust uses it more like the latter.

Sorry for the last-minute comment. I don't want to bikeshed naming, but I don't think this is a bikeshed concern. I think this is likely to produce confusion.

CDirkx commented:
Maybe `Unsupported`? Like the `unsupported` function and `std::sys::unsupported`.

joshtriplett commented:
@CDirkx Yeah, that seems fine. That's different enough that it doesn't *directly* sound like a direct mapping of a specific different error code. The rest could be handled via documentation.

Dylan-DPC commented:
r? @joshtriplett

CDirkx commented:
I renamed `NotSupported` to `Unsupported` and added an extra bit of documentation: "This means that the operation can never succeed."

joshtriplett commented:
@bors r+

bors commented:
:pushpin: Commit 0c3303e140d1254dc5c0b96d17ce9b5e689341a4 has been approved by `joshtriplett`

<!-- @bors r=joshtriplett 0c3303e140d1254dc5c0b96d17ce9b5e689341a4 -->
<!-- homu: {"type":"Approved","sha":"0c3303e140d1254dc5c0b96d17ce9b5e689341a4","approver":"joshtriplett"} -->

bors commented:
:hourglass: Testing commit 0c3303e140d1254dc5c0b96d17ce9b5e689341a4 with merge ba953fbac71474459bb537b4138bd53885caf252...
<!-- homu: {"type":"BuildStarted","head_sha":"0c3303e140d1254dc5c0b96d17ce9b5e689341a4","merge_sha":"ba953fbac71474459bb537b4138bd53885caf252"} -->

rust-log-analyzer commented:
The job **`dist-i586-gnu-i586-i686-musl`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/2146096843) [(plain)](https://github.com/rust-lang-ci/rust/commit/ba953fbac71474459bb537b4138bd53885caf252/checks/2146096843/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test sync::mpsc::tests::stress_recv_timeout_two_threads ... ok

failures:

---- fs::tests::metadata_access_times stdout ----
thread '<unnamed>' panicked at 'creation time must be always supported or not supported: Err(Custom { kind: Unsupported, error: "creation time is not available on this platform currently" }) Err(Custom { kind: Unsupported, error: "creation time is not available on this platform currently" })', library/std/src/fs/tests.rs:1334:17

failures:
    fs::tests::metadata_access_times


test result: FAILED. 852 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 10.33s

error: test failed, to rerun pass '-p std --lib'


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "i686-unknown-linux-musl" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "std" "--"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --host= --target i586-unknown-linux-gnu,i686-unknown-linux-musl
Build completed unsuccessfully in 0:24:45
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2146226797)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/2146226797","builder_name":"checks-actions"} -->

m-ou-se commented on library/std/src/io/error.rs:
```diff
@@ -178,6 +178,12 @@ pub enum ErrorKind {
+    /// This operation is unsupported on this platform.
+    ///
+    /// This means that the operation can never succeed.
+    #[stable(feature = "unsupported_error", since = "1.52.0")]
```
Nightly is at 1.53.0 now:
```suggestion
    #[stable(feature = "unsupported_error", since = "1.53.0")]
```

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2202768814) [(plain)](https://github.com/rust-lang/rust/commit/697db2671ee0222b31bb5f4a87b98af6055f761a/checks/2202768814/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/std/src/sys/windows/fs.rs at line 802:
 
 #[cfg(target_vendor = "uwp")]
 pub fn link(_original: &Path, _link: &Path) -> io::Result<()> {
-    return Err(io::Error::new_const(io::ErrorKind::Unsupported, &"hard link are not supported on UWP"));
+    return Err(io::Error::new_const(
+        io::ErrorKind::Unsupported,
+        &"hard link are not supported on UWP",
 }
 
 
 pub fn stat(path: &Path) -> io::Result<FileAttr> {
Diff in /checkout/library/std/src/sys/unix/l4re.rs at line 1:
 macro_rules! unimpl {
     () => {
-        return Err(io::Error::new_const(io::ErrorKind::Unsupported, &"No networking available on L4Re."));
+        return Err(io::Error::new_const(
+            io::ErrorKind::Unsupported,
+            &"No networking available on L4Re.",
     };
 }
 
 
Diff in /checkout/library/std/src/sys/unsupported/common.rs at line 18:
 
 
 pub fn unsupported_err() -> std_io::Error {
-    std_io::Error::new_const(std_io::ErrorKind::Unsupported, &"operation not supported on this platform")
+    std_io::Error::new_const(
+        std_io::ErrorKind::Unsupported,
+        &"operation not supported on this platform",
 }
 
 
 pub fn decode_error_kind(_code: i32) -> crate::io::ErrorKind {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/sys/windows/stdio.rs" "/checkout/library/std/src/sys/windows/alloc.rs" "/checkout/library/std/src/sys/windows/net.rs" "/checkout/library/std/src/sys/windows/memchr.rs" "/checkout/library/std/src/sys/windows/fs.rs" "/checkout/library/std/src/sys/windows/thread.rs" "/checkout/library/std/src/sys/windows/args.rs" "/checkout/library/std/src/sys/windows/process.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:14
```

</details>

CDirkx commented:
Merge conflicts have been resolved.

joshtriplett commented:
@bors r+

bors commented:
:pushpin: Commit 4f7d87ea987bb4d18a8028e64cb0b459a246f7e2 has been approved by `joshtriplett`

<!-- @bors r=joshtriplett 4f7d87ea987bb4d18a8028e64cb0b459a246f7e2 -->
<!-- homu: {"type":"Approved","sha":"4f7d87ea987bb4d18a8028e64cb0b459a246f7e2","approver":"joshtriplett"} -->

bors commented:
:hourglass: Testing commit 4f7d87ea987bb4d18a8028e64cb0b459a246f7e2 with merge f7fdc3a3393ad26af1dbdff18e8ee3772479580a...
<!-- homu: {"type":"BuildStarted","head_sha":"4f7d87ea987bb4d18a8028e64cb0b459a246f7e2","merge_sha":"f7fdc3a3393ad26af1dbdff18e8ee3772479580a"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/2213858719) [(plain)](https://github.com/rust-lang-ci/rust/commit/f7fdc3a3393ad26af1dbdff18e8ee3772479580a/checks/2213858719/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
error: tests/compile-fail/unaligned_pointers/reference_to_packed.rs:18: unexpected warning: '18:26: 18:32: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!'

error: 2 unexpected errors found, 0 expected errors not found
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/miri" "tests/compile-fail/unaligned_pointers/reference_to_packed.rs" "-L" "/tmp/compiletestHYHFgU" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/compiletestHYHFgU/unaligned_pointers/reference_to_packed.stage-id" "-A" "unused" "--edition" "2018" "-Astable-features" "--sysroot" "/home/user/.cache/miri/HOST" "-Zmiri-disable-validation" "-Zmiri-disable-stacked-borrows" "-L" "/tmp/compiletestHYHFgU/unaligned_pointers/reference_to_packed.stage-id.aux"
   1: compiletest_rs::runtest::TestCx::run_cfail_test
   2: compiletest_rs::runtest::run
   3: core::ops::function::FnOnce::call_once{{vtable.shim}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
---
Cloning into 'rust-toolstate'...
{"cargo-miri":"test-fail","rustbook":"test-fail","rls":"build-fail","nomicon":"test-pass","embedded-book":"test-pass","miri":"test-fail","rust-by-example":"test-pass","reference":"test-pass","book":"test-pass","rustfmt":"build-fail","edition-guide":"test-pass"}[master 1b8d68e] (linux CI update)
 1 file changed, 1 insertion(+)
To https://github.com/rust-lang-nursery/rust-toolstate
   6b7b1a3..1b8d68e  master -> master
Build completed successfully in 0:00:03
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
[TIMING] Assemble { target_compiler: Compiler { stage: 0, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } } -- 0.000
[TIMING] StartupObjects { compiler: Compiler { stage: 0, host: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } }, target: TargetSelection { triple: "x86_64-unknown-linux-gnu", file: None } } -- 0.000
---

---- compile_test stdout ----
diff of stderr:

 error: wildcard match will also match any future added variants
   --> $DIR/wildcard_enum_match_arm.rs:39:9
    |
 LL |         _ => eprintln!("Not red"),
    |         ^ help: try this: `Color::Green | Color::Blue | Color::Rgb(..) | Color::Cyan`
 note: the lint level is defined here
   --> $DIR/wildcard_enum_match_arm.rs:3:9
    |
    |
 LL | #![deny(clippy::wildcard_enum_match_arm)]
 
 
 error: wildcard match will also match any future added variants
   --> $DIR/wildcard_enum_match_arm.rs:43:9
    |
 LL |         _not_red => eprintln!("Not red"),
    |         ^^^^^^^^ help: try this: `_not_red @ Color::Green | _not_red @ Color::Blue | _not_red @ Color::Rgb(..) | _not_red @ Color::Cyan`
 
 error: wildcard match will also match any future added variants
   --> $DIR/wildcard_enum_match_arm.rs:47:9
    |
 LL |         not_red => format!("{:?}", not_red),
    |         ^^^^^^^ help: try this: `not_red @ Color::Green | not_red @ Color::Blue | not_red @ Color::Rgb(..) | not_red @ Color::Cyan`
 
 error: wildcard match will also match any future added variants
   --> $DIR/wildcard_enum_match_arm.rs:63:9
    |
 LL |         _ => "No red",
    |         ^ help: try this: `Color::Red | Color::Green | Color::Blue | Color::Rgb(..) | Color::Cyan`
 
 
 error: wildcard matches known variants and will also match future added variants
   --> $DIR/wildcard_enum_match_arm.rs:80:9
 LL |         _ => {},
 LL |         _ => {},
-   |         ^ help: try this: `ErrorKind::PermissionDenied | ErrorKind::ConnectionRefused | ErrorKind::ConnectionReset | ErrorKind::ConnectionAborted | ErrorKind::NotConnected | ErrorKind::AddrInUse | ErrorKind::AddrNotAvailable | ErrorKind::BrokenPipe | ErrorKind::AlreadyExists | ErrorKind::WouldBlock | ErrorKind::InvalidInput | ErrorKind::InvalidData | ErrorKind::TimedOut | ErrorKind::WriteZero | ErrorKind::Interrupted | ErrorKind::Other | ErrorKind::UnexpectedEof | _`
+   |         ^ help: try this: `ErrorKind::PermissionDenied | ErrorKind::ConnectionRefused | ErrorKind::ConnectionReset | ErrorKind::ConnectionAborted | ErrorKind::NotConnected | ErrorKind::AddrInUse | ErrorKind::AddrNotAvailable | ErrorKind::BrokenPipe | ErrorKind::AlreadyExists | ErrorKind::WouldBlock | ErrorKind::InvalidInput | ErrorKind::InvalidData | ErrorKind::TimedOut | ErrorKind::WriteZero | ErrorKind::Interrupted | ErrorKind::Other | ErrorKind::UnexpectedEof | ErrorKind::Unsupported | _`
-error: aborting due to 5 previous errors
-error: aborting due to 5 previous errors
+error: wildcard matches known variants and will also match future added variants
+  --> $DIR/wildcard_enum_match_arm.rs:101:9
+LL |         _ => {},
+LL |         _ => {},
+   |         ^ help: try this: `ErrorKind::Unsupported | _`
+error: aborting due to 6 previous errors
 
 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/build/clippy-19ca3b17afe90112/out/test_build_base/wildcard_enum_match_arm.stage-id.stderr

 // run-rustfix
 
 
 #![deny(clippy::wildcard_enum_match_arm)]
 #![allow(
     unreachable_code,
     unused_variables,
     clippy::single_match,
     clippy::single_match,
     clippy::wildcard_in_or_patterns,
     clippy::unnested_or_patterns,
     clippy::diverging_sub_expression
 
 use std::io::ErrorKind;
 
 
 #[derive(Clone, Copy, Debug, Eq, PartialEq)]
 enum Color {
     Red,
     Green,
     Blue,
     Rgb(u8, u8, u8),
     Cyan,
 
 impl Color {
 impl Color {
     fn is_monochrome(self) -> bool {
         match self {
             Color::Red | Color::Green | Color::Blue => true,
             Color::Rgb(r, g, b) => r | g == 0 || r | b == 0 || g | b == 0,
             Color::Cyan => false,
     }
 }
 
 fn main() {
 fn main() {
     let color = Color::Rgb(0, 0, 127);
     match color {
         Color::Red => println!("Red"),
         Color::Green | Color::Blue | Color::Rgb(..) | Color::Cyan => eprintln!("Not red"),
     match color {
     match color {
         Color::Red => println!("Red"),
         _not_red @ Color::Green | _not_red @ Color::Blue | _not_red @ Color::Rgb(..) | _not_red @ Color::Cyan => eprintln!("Not red"),
     let _str = match color {
     let _str = match color {
         Color::Red => "Red".to_owned(),
         not_red @ Color::Green | not_red @ Color::Blue | not_red @ Color::Rgb(..) | not_red @ Color::Cyan => format!("{:?}", not_red),
     match color {
     match color {
         Color::Red => {},
         Color::Green => {},
         Color::Blue => {},
         Color::Cyan => {},
         c if c.is_monochrome() => {},
         Color::Rgb(_, _, _) => {},
     let _str = match color {
     let _str = match color {
         Color::Red => "Red",
         c @ Color::Green | c @ Color::Blue | c @ Color::Rgb(_, _, _) | c @ Color::Cyan => "Not red",
     match color {
     match color {
         Color::Rgb(r, _, _) if r > 0 => "Some red",
         Color::Red | Color::Green | Color::Blue | Color::Rgb(..) | Color::Cyan => "No red",
     match color {
     match color {
         Color::Red | Color::Green | Color::Blue | Color::Cyan => {},
         Color::Rgb(..) => {},
     };
     let x: u8 = unimplemented!();
     match x {
         0 => {},
         140 => {},
         _ => {},
     };
     // We need to use an enum not defined in this test because non_exhaustive is ignored for the
     // purposes of dead code analysis within a crate.
     let error_kind = ErrorKind::NotFound;
     match error_kind {
         ErrorKind::NotFound => {},
-        ErrorKind::PermissionDenied | ErrorKind::ConnectionRefused | ErrorKind::ConnectionReset | ErrorKind::ConnectionAborted | ErrorKind::NotConnected | ErrorKind::AddrInUse | ErrorKind::AddrNotAvailable | ErrorKind::BrokenPipe | ErrorKind::AlreadyExists | ErrorKind::WouldBlock | ErrorKind::InvalidInput | ErrorKind::InvalidData | ErrorKind::TimedOut | ErrorKind::WriteZero | ErrorKind::Interrupted | ErrorKind::Other | ErrorKind::UnexpectedEof | _ => {},
+        ErrorKind::PermissionDenied | ErrorKind::ConnectionRefused | ErrorKind::ConnectionReset | ErrorKind::ConnectionAborted | ErrorKind::NotConnected | ErrorKind::AddrInUse | ErrorKind::AddrNotAvailable | ErrorKind::BrokenPipe | ErrorKind::AlreadyExists | ErrorKind::WouldBlock | ErrorKind::InvalidInput | ErrorKind::InvalidData | ErrorKind::TimedOut | ErrorKind::WriteZero | ErrorKind::Interrupted | ErrorKind::Other | ErrorKind::UnexpectedEof | ErrorKind::Unsupported | _ => {},
     match error_kind {
         ErrorKind::NotFound => {},
         ErrorKind::NotFound => {},
         ErrorKind::PermissionDenied => {},
         ErrorKind::ConnectionRefused => {},
         ErrorKind::ConnectionReset => {},
         ErrorKind::ConnectionAborted => {},
         ErrorKind::NotConnected => {},
         ErrorKind::AddrInUse => {},
         ErrorKind::AddrNotAvailable => {},
         ErrorKind::BrokenPipe => {},
         ErrorKind::AlreadyExists => {},
         ErrorKind::WouldBlock => {},
         ErrorKind::InvalidInput => {},
         ErrorKind::InvalidData => {},
         ErrorKind::TimedOut => {},
         ErrorKind::WriteZero => {},
         ErrorKind::Interrupted => {},
         ErrorKind::Other => {},
         ErrorKind::UnexpectedEof => {},
-        _ => {},
+        ErrorKind::Unsupported | _ => {},
 }
 

The actual fixed differed from the expected fixed.
The actual fixed differed from the expected fixed.
Actual fixed saved to /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/build/clippy-19ca3b17afe90112/out/test_build_base/wildcard_enum_match_arm.stage-id.fixed
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args wildcard_enum_match_arm.rs`
error: 2 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/clippy-driver" "tests/ui/wildcard_enum_match_arm.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/build/clippy-19ca3b17afe90112/out/test_build_base" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/build/clippy-19ca3b17afe90112/out/test_build_base/wildcard_enum_match_arm.stage-id" "-A" "unused" "--emit=metadata" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "-Dwarnings" "-Zui-testing" "--extern" "syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-3e6c923d343daab8.rlib" "--extern" "clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-2a79111c7ffd0851.rlib" "--extern" "regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-4cd0b4140eeb9c8d.rlib" "--extern" "serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-3ddf9f6cc09731de.rlib" "--extern" "quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-9816c8a0844f9255.rlib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/build/clippy-19ca3b17afe90112/out/test_build_base/wildcard_enum_match_arm.stage-id.aux"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
{"message":"wildcard match will also match any future added variants","code":{"code":"clippy::wildcard_enum_match_arm","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":788,"byte_end":789,"line_start":39,"line_end":39,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"        _ => eprintln!(\"Not red\"),","highlight_start":9,"highlight_end":10}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"the lint level is defined here","code":null,"level":"note","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":24,"byte_end":55,"line_start":3,"line_end":3,"column_start":9,"column_end":40,"is_primary":true,"text":[{"text":"#![deny(clippy::wildcard_enum_match_arm)]","highlight_start":9,"highlight_end":40}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[],"rendered":null},{"message":"try this","code":null,"level":"help","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":788,"byte_end":789,"line_start":39,"line_end":39,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"        _ => eprintln!(\"Not red\"),","highlight_start":9,"highlight_end":10}],"label":null,"suggested_replacement":"Color::Green | Color::Blue | Color::Rgb(..) | Color::Cyan","suggestion_applicability":"MaybeIncorrect","expansion":null}],"children":[],"rendered":null}],"rendered":"error: wildcard match will also match any future added variants\n  --> tests/ui/wildcard_enum_match_arm.rs:39:9\n   |\nLL |         _ => eprintln!(\"Not red\"),\n   |         ^ help: try this: `Color::Green | Color::Blue | Color::Rgb(..) | Color::Cyan`\n   |\nnote: the lint level is defined here\n  --> tests/ui/wildcard_enum_match_arm.rs:3:9\n   |\nLL | #![deny(clippy::wildcard_enum_match_arm)]\n   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n\n"}
{"message":"wildcard match will also match any future added variants","code":{"code":"clippy::wildcard_enum_match_arm","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":887,"byte_end":895,"line_start":43,"line_end":43,"column_start":9,"column_end":17,"is_primary":true,"text":[{"text":"        _not_red => eprintln!(\"Not red\"),","highlight_start":9,"highlight_end":17}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"try this","code":null,"level":"help","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":887,"byte_end":895,"line_start":43,"line_end":43,"column_start":9,"column_end":17,"is_primary":true,"text":[{"text":"        _not_red => eprintln!(\"Not red\"),","highlight_start":9,"highlight_end":17}],"label":null,"suggested_replacement":"_not_red @ Color::Green | _not_red @ Color::Blue | _not_red @ Color::Rgb(..) | _not_red @ Color::Cyan","suggestion_applicability":"MaybeIncorrect","expansion":null}],"children":[],"rendered":null}],"rendered":"error: wildcard match will also match any future added variants\n  --> tests/ui/wildcard_enum_match_arm.rs:43:9\n   |\nLL |         _not_red => eprintln!(\"Not red\"),\n   |         ^^^^^^^^ help: try this: `_not_red @ Color::Green | _not_red @ Color::Blue | _not_red @ Color::Rgb(..) | _not_red @ Color::Cyan`\n\n"}
{"message":"wildcard match will also match any future added variants","code":{"code":"clippy::wildcard_enum_match_arm","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":1005,"byte_end":1012,"line_start":47,"line_end":47,"column_start":9,"column_end":16,"is_primary":true,"text":[{"text":"        not_red => format!(\"{:?}\", not_red),","highlight_start":9,"highlight_end":16}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"try this","code":null,"level":"help","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":1005,"byte_end":1012,"line_start":47,"line_end":47,"column_start":9,"column_end":16,"is_primary":true,"text":[{"text":"        not_red => format!(\"{:?}\", not_red),","highlight_start":9,"highlight_end":16}],"label":null,"suggested_replacement":"not_red @ Color::Green | not_red @ Color::Blue | not_red @ Color::Rgb(..) | not_red @ Color::Cyan","suggestion_applicability":"MaybeIncorrect","expansion":null}],"children":[],"rendered":null}],"rendered":"error: wildcard match will also match any future added variants\n  --> tests/ui/wildcard_enum_match_arm.rs:47:9\n   |\nLL |         not_red => format!(\"{:?}\", not_red),\n   |         ^^^^^^^ help: try this: `not_red @ Color::Green | not_red @ Color::Blue | not_red @ Color::Rgb(..) | not_red @ Color::Cyan`\n\n"}
{"message":"wildcard match will also match any future added variants","code":{"code":"clippy::wildcard_enum_match_arm","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":1499,"byte_end":1500,"line_start":63,"line_end":63,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"        _ => \"No red\",","highlight_start":9,"highlight_end":10}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"try this","code":null,"level":"help","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":1499,"byte_end":1500,"line_start":63,"line_end":63,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"        _ => \"No red\",","highlight_start":9,"highlight_end":10}],"label":null,"suggested_replacement":"Color::Red | Color::Green | Color::Blue | Color::Rgb(..) | Color::Cyan","suggestion_applicability":"MaybeIncorrect","expansion":null}],"children":[],"rendered":null}],"rendered":"error: wildcard match will also match any future added variants\n  --> tests/ui/wildcard_enum_match_arm.rs:63:9\n   |\nLL |         _ => \"No red\",\n   |         ^ help: try this: `Color::Red | Color::Green | Color::Blue | Color::Rgb(..) | Color::Cyan`\n\n"}
{"message":"wildcard matches known variants and will also match future added variants","code":{"code":"clippy::wildcard_enum_match_arm","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":2012,"byte_end":2013,"line_start":80,"line_end":80,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"        _ => {},","highlight_start":9,"highlight_end":10}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"try this","code":null,"level":"help","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":2012,"byte_end":2013,"line_start":80,"line_end":80,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"        _ => {},","highlight_start":9,"highlight_end":10}],"label":null,"suggested_replacement":"ErrorKind::PermissionDenied | ErrorKind::ConnectionRefused | ErrorKind::ConnectionReset | ErrorKind::ConnectionAborted | ErrorKind::NotConnected | ErrorKind::AddrInUse | ErrorKind::AddrNotAvailable | ErrorKind::BrokenPipe | ErrorKind::AlreadyExists | ErrorKind::WouldBlock | ErrorKind::InvalidInput | ErrorKind::InvalidData | ErrorKind::TimedOut | ErrorKind::WriteZero | ErrorKind::Interrupted | ErrorKind::Other | ErrorKind::UnexpectedEof | ErrorKind::Unsupported | _","suggestion_applicability":"MaybeIncorrect","expansion":null}],"children":[],"rendered":null}],"rendered":"error: wildcard matches known variants and will also match future added variants\n  --> tests/ui/wildcard_enum_match_arm.rs:80:9\n   |\nLL |         _ => {},\n   |         ^ help: try this: `ErrorKind::PermissionDenied | ErrorKind::ConnectionRefused | ErrorKind::ConnectionReset | ErrorKind::ConnectionAborted | ErrorKind::NotConnected | ErrorKind::AddrInUse | ErrorKind::AddrNotAvailable | ErrorKind::BrokenPipe | ErrorKind::AlreadyExists | ErrorKind::WouldBlock | ErrorKind::InvalidInput | ErrorKind::InvalidData | ErrorKind::TimedOut | ErrorKind::WriteZero | ErrorKind::Interrupted | ErrorKind::Other | ErrorKind::UnexpectedEof | ErrorKind::Unsupported | _`\n\n"}
{"message":"wildcard matches known variants and will also match future added variants","code":{"code":"clippy::wildcard_enum_match_arm","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":2756,"byte_end":2757,"line_start":101,"line_end":101,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"        _ => {},","highlight_start":9,"highlight_end":10}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"try this","code":null,"level":"help","spans":[{"file_name":"tests/ui/wildcard_enum_match_arm.rs","byte_start":2756,"byte_end":2757,"line_start":101,"line_end":101,"column_start":9,"column_end":10,"is_primary":true,"text":[{"text":"        _ => {},","highlight_start":9,"highlight_end":10}],"label":null,"suggested_replacement":"ErrorKind::Unsupported | _","suggestion_applicability":"MaybeIncorrect","expansion":null}],"children":[],"rendered":null}],"rendered":"error: wildcard matches known variants and will also match future added variants\n  --> tests/ui/wildcard_enum_match_arm.rs:101:9\n   |\nLL |         _ => {},\n   |         ^ help: try this: `ErrorKind::Unsupported | _`\n\n"}

------------------------------------------

thread 'compile_test' panicked at 'Some tests failed', /cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.6.0/src/lib.rs:105:22
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2213958101)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/2213958101","builder_name":"checks-actions"} -->

CDirkx commented:
I fixed the clippy test that uses `ErrorKind`, but I'm not sure how the failure in `compile-fail/unaligned_pointers/reference_to_packed.rs` is related?

jyn514 commented:
@CDirkx hmm, try pushing your clippy fix and rebasing over master? I agree the compile-fail clippy test looks unrelated.

CDirkx commented:
Rebased and fixed the clippy error, should be ready to try merging again.

jyn514 commented:
@bors r=joshtriplett

bors commented:
:pushpin: Commit 5b5afaefff5bdfc01911b9013318145466abec68 has been approved by `joshtriplett`

<!-- @bors r=joshtriplett 5b5afaefff5bdfc01911b9013318145466abec68 -->
<!-- homu: {"type":"Approved","sha":"5b5afaefff5bdfc01911b9013318145466abec68","approver":"joshtriplett"} -->

bors commented:
:hourglass: Testing commit 5b5afaefff5bdfc01911b9013318145466abec68 with merge 0052f444a7935147d31baa80099dbdfb1140255d...
<!-- homu: {"type":"BuildStarted","head_sha":"5b5afaefff5bdfc01911b9013318145466abec68","merge_sha":"0052f444a7935147d31baa80099dbdfb1140255d"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-apple-alt`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/2375480070) [(plain)](https://github.com/rust-lang-ci/rust/commit/0052f444a7935147d31baa80099dbdfb1140255d/checks/2375480070/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
      Memory: 14 GB
      Boot ROM Version: VMW71.00V.13989454.B64.1906190538
      Apple ROM Info: [MS_VM_CERT/SHA1/27d66596a61c48dd3dc7216fd715126e33f59ae7]Welcome to the Virtual Machine
      SMC Version (system): 2.8f0
      Serial Number (system): VMdSnO8Wr7RX

hw.ncpu: 3
hw.byteorder: 1234
hw.memsize: 15032385536
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2375730751)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/2375730751","builder_name":"checks-actions"} -->

jyn514 commented:
```
error: failed to download from `https://crates.io/api/v1/crates/minifier/0.0.39/download`
failed to run: /Users/runner/work/rust/rust/build/bootstrap/debug/bootstrap dist

Caused by:
  failed to get 200 response from `https://crates.io/api/v1/crates/minifier/0.0.39/download`, got 503
```
@bors retry

bors commented:
:hourglass: Testing commit 5b5afaefff5bdfc01911b9013318145466abec68 with merge 5a4ab26459a1ccf17ef5bb4c841d3ae5517b2890...
<!-- homu: {"type":"BuildStarted","head_sha":"5b5afaefff5bdfc01911b9013318145466abec68","merge_sha":"5a4ab26459a1ccf17ef5bb4c841d3ae5517b2890"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2376287698)
Approved by: joshtriplett
Pushing 5a4ab26459a1ccf17ef5bb4c841d3ae5517b2890 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"joshtriplett","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2376287698"},"merge_sha":"5a4ab26459a1ccf17ef5bb4c841d3ae5517b2890"} -->
